### ISSUE TITLE
Add notarization to unified mac client

### ISSUE THREAD

**Issue by karanbirsingh** [microsoft/accessibility-insights-web#2154]
## Describe the bug

Starting Feb 3, the startup workflow on MacOS Catalina is different for unnotarized applications: 
- https://developer.apple.com/news/?id=12232019a
- https://support.apple.com/en-gb/HT202491

We need to check how the unified mac client runs on Catalina, and consider notarizing the app (in addition to existing signing) to remove extra end-user work.

**Comment by ferBonnin** [comment]
Per conversation with Peter, let's do this.

**Comment by jalkire** [comment]
Added the [ESRP notarization task to our signed build](https://github.com/microsoft/accessibility-insights-web/blob/144fbd6b92d02c202907c375dcaa41e5d5b2bd61/pipeline/unified/channel/sign-release-package.yaml#L41). Resulted in [an error with no details](https://dev.azure.com/accessibility-insights-private/Accessibility%20Insights%20(private)/_build/results?buildId=12115&view=logs&j=c4b2f6bc-f488-552b-1554-f6c085ffa0db&t=fc4d40c4-9764-59f7-9a06-22984310b4b5). Unassigning myself to start on feature work.

**Comment by dbjorge** [comment]
Picking this up where John left off. Next few items I'm going to try, based on [Apple's notarization docs](https://developer.apple.com/documentation/xcode/notarizing_macos_software_before_distribution):

* [ ] Bring John's branch up to date with master
* [ ] Verify that electron-builder's output has the [hardened runtime](https://help.apple.com/xcode/mac/current/#/devf87a2ac8f) enabled (electron-builder's [mac config docs](https://www.electron.build/configuration/mac.html) suggest that `hardenedRuntime=true` is the default configuration, but I want to double check because we've seen misleading behavior from electron-builder config before)
* [ ] Verify that electron-builder's output has an entitlements file, and/or try setting one explicitly. When doing mac codesigning electron-builder defers to `electron-osx-sign`, which has a default entitlements file, but since we're doing our own signing maybe we aren't getting that benefit?
* [ ] Verify the bundle ID is getting set correctly at build time
* [ ] Double check our ESRP parameters
* [ ] Reach out to ESRP support
* [ ] Find examples of other teams using ESRP notarization and see if they're doing anything exciting

**Comment by dbjorge** [comment]
Followed the [VS Code team's example](https://github.com/microsoft/vscode/blob/227d533937b17f54bae09f94083bf6304e6262e1/build/azure-pipelines/darwin/product-build-darwin.yml); the key realization is that it's the codesigning process that "enables" the hardened runtime in an electron build process, so I needed to add a parameter to enable it to the signing (not notarizing) step of the original ESRP invocation. There were a few other minor changes I made during testing that I kept, but I think that was the key one.

* [Build #Signing-478](https://dev.azure.com/accessibility-insights-private/Accessibility%20Insights%20(private)/_build/results?buildId=13430&view=artifacts&type=publishedArtifacts) with first examples of notarize step succeeding

@waabid is testing the resulting artifacts now as I write this, fingers crossed :)

**Comment by dbjorge** [comment]
Wahaj found that the app crashes on startup. The relevant snippet of error is:

```
Process:               Accessibility Insights for Android - Insider [98982]
Path:                  /Applications/Accessibility Insights for Android - Insider.app/Contents/MacOS/Accessibility Insights for Android - Insider
Identifier:            com.microsoft.accessibilityinsights.unified.insider
Version:               2020.803.9 (2020.803.9.2020.803.9)
Code Type:             X86-64 (Native)
Parent Process:        ??? [1]
Responsible:           Accessibility Insights for Android - Insider [98982]
User ID:               501

Date/Time:             2020-08-03 15:10:51.195 -0700
OS Version:            Mac OS X 10.15.5 (19F101)
Report Version:        12
Bridge OS Version:     4.5 (17P5300)
Anonymous UUID:        D58B1136-DAB4-4384-860F-9D25A6897970

Sleep/Wake UUID:       7C8CC498-A472-4469-8CC4-EF94FA722520

Time Awake Since Boot: 59000 seconds
Time Since Wake:       4900 seconds

System Integrity Protection: enabled

Crashed Thread:        0  Dispatch queue: com.apple.main-thread

Exception Type:        EXC_BAD_ACCESS (Code Signature Invalid)
Exception Codes:       0x0000000000000032, 0x0000316972d02040
Exception Note:        EXC_CORPSE_NOTIFY

Termination Reason:    Namespace CODESIGNING, Code 0x2
```

Related issues:
* https://github.com/electron-userland/electron-builder/issues/3989
* https://github.com/electron-userland/electron-builder/issues/4040#issuecomment-615839544

Karan suggested a second example to look at (Bot Framework):
* [PR where they added the notarization yaml logic](https://github.com/microsoft/BotFramework-Emulator/pull/2135/files)
* [Non-YAML build where they perform some of the pre-signing steps](https://fuselabs.visualstudio.com/BotFramework-Emulator/_apps/hub/ms.vss-ciworkflow.build-ci-hub?_a=edit-build-definition&id=394)

They're using the same basic strategy as VS Code (performing a pre-ESRP-signing step that uses the `codesign` tool on the dev box to assign an enlistments file and a hardening config, *then* sending that to ESRP. Like VS Code, they're signing a .app file (not a .dmg like we're doing), so I can only hope it's a similar process.

**Comment by dbjorge** [comment]
Talked to the ESRP team today and confirmed with them that following the examples/guidance from the VS Code and BotFramework-Emulator repos is correct. At a high level, the process is:

* ESRP folks add our build/release admin group as a "derived" member of the general Microsoft account for the Apple Developer Program
* We generate a (derived) Apple Developer ID Certificate from the Apple Developer Program portal
* We store the certificate (base64-encoded) and its password in the KeyVault instance with the rest of our publishing secrets
* Our release/signing builds consume that certificate/password via a KeyVault-linked variable group and add it to a `.keychain` file on the build agent
* We use `codesign` (either directly like BotEmulator-Framework, via `electron-osx-sign` like VS Code, or via `electron-builder` (which uses `electron-osx-sign`) to apply the "hardened runtime" option and the entitlements `.plist`s (for our main .app and all its helpers). This must happen after `electron-builder` produces the release directory structure but before the structure is packed into `.dmg`/`.zip` files
  - This does result in signing the `.app`s with our derived Developer ID cert; the signature doesn't matter so much, this is mostly a side effect of how you set the entitlements
* We pack the `.zip`/`.dmg` as we do now
* We send it to ESRP for codesigning (with a special hardened runtime option)
  - ESRP knows how to maintain the existing entitlements we applied earlier as part of the "official" signature
* We send it to ESRP again for notarization

**Comment by dbjorge** [comment]
We've now worked with the ESRP folks and provisioned an Apple developer account that can generate the derived cert we'll need. I've also hooked up placeholder secrets in our usual keyvault and piped them into the signing build. The next steps all require an actual mac for testing/validation:

* Generating the certificate we need requires using a mac to create a "certificate signing request" and uploading that to the Apple Developer Portal.
* It's not clear which strategy for using the cert will be easiest (we may be able to just use electron-builder directly by passing it the appropriate CSC_* environment variables, but I haven't tested this and there's probably a reason BotFramework and VS Code aren't doing that). If that doesn't work, we'll need to iterate on splitting the electron-builder step in our build into separate "prepare folder structure" and "package into dmg" steps, so we can insert the codesign step in between them.

I've asked Peter to have a mac provisioned for me; moving this feature back to ready for work until I receive it or someone else with a Mac can pick it up, rolling onto web compliance feature in meantime.
